axiom/tokens_integration_test.go

```diff
@@ -0,0 +1,91 @@
+//go:build integration
+
+package axiom_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/suite"
+
+	"github.com/axiomhq/axiom-go/axiom"
+)
+
+// TokensTestSuite tests all methods of the Axiom Tokens API against a
+// live deployment.
+type TokensTestSuite struct {
+	IntegrationTestSuite
+
+	apiToken *axiom.APIToken
+}
+
+func TestTokensTestSuite(t *testing.T) {
+	suite.Run(t, new(TokensTestSuite))
+}
+
+func (s *TokensTestSuite) SetupSuite() {
+	s.IntegrationTestSuite.SetupSuite()
+}
+
+func (s *TokensTestSuite) TearDownSuite() {
+	// Teardown routines use their own context to avoid not being run at all
+	// when the suite gets cancelled or times out.
+	ctx, cancel := context.WithTimeout(context.Background(), 15*time.Second)
+	defer cancel()
+
+	err := s.client.Tokens.Delete(ctx, s.apiToken.ID)
+	s.NoError(err)
+
+	s.IntegrationTestSuite.TearDownSuite()
+}
+
+func (s *TokensTestSuite) Test() {
+	createdToken, err := s.client.Tokens.Create(s.suiteCtx, axiom.CreateTokenRequest{
+		Name:      "Test token",
+		ExpiresAt: time.Now().Add(24 * time.Hour),
+		DatasetCapabilities: map[string]axiom.DatasetCapabilities{
+			"*": {Ingest: []string{"create"}}},
+		OrganisationCapabilities: axiom.OrganisationCapabilities{
+			Users: []string{"create", "read", "update", "delete"},
+		}})
+	s.Require().NoError(err)
+	s.Require().NotNil(createdToken)
+
+	s.apiToken = createdToken.AsAPIToken()
```

Comment:
If creation of the token fails for some reason, the test suit will call `TearDownSuite` which in turn calls `s.apiToken.ID` where `apiToken` is `nil`. This will cause a panic.

Please move the creation logic into `SetupTest`, just like the teardown/deletion lives in `TearDown...`. Check the integrations tests for the annotations or datasets service for reference.

Reply:
Thanks have moved it to SetupTest